### ISSUE TITLE
Elasticsearch 8 composer version constraint is not allowing latest versions (PHP 8.3 related)

### ISSUE THREAD

**Issue by hostep** [magento/magento2#38416]
### Preconditions and environment

- Magento version 2.4-develop

### Steps to reproduce

Try to update `elasticsearch/elasticsearch` package to latest version (8.12.0 at the moment of writing):
```
$ composer update elasticsearch/elasticsearch:8.12.0
The temporary constraint "8.12.0" for "elasticsearch/elasticsearch" must be a subset of the constraint in your composer.json (~7.17.0 || ~8.5.0)
Run `composer require elasticsearch/elasticsearch` or `composer require elasticsearch/elasticsearch:8.12.0` instead to replace the constraint
```

### Expected result

Should be possible to install latest version of version 8 of `elasticsearch/elasticsearch`

### Actual result

Not possible to install latest version of version 8 of `elasticsearch/elasticsearch`, due to constraint `~8.5.0` being used by Magento codebase

### Additional information

Version 8.12.0 of `elasticsearch/elasticsearch` was tested with PHP 8.3 as they claim here: https://github.com/elastic/elasticsearch-php/releases/tag/v8.12.0
So it might mean that older versions of this package might not be 100% compatible with PHP 8.3 yet.

Currently magento supports elasticsearch till 8.7, but version constraint is set to 8.5.

### Triage and priority

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by engcom-November** [comment]
Hello @hostep,

Thank you for the report!

Verified this issue on 2.4-develop.
With PHP 8.3, when trying to update elasticsearch via compser we got the below error:

<img width="1025" alt="image" src="https://github.com/magento/magento2/assets/60198231/2e8ee46b-1981-4f45-a365-64800af125d2">

Hence confirming the issue.

Thank you.

**Comment by engcom-November** [comment]
Hello @hostep,

As mentioned in the system requirements [here](https://experienceleague.adobe.com/docs/commerce-operations/installation-guide/system-requirements.html?lang=en), magento currently supports elasticsearch 8.7.
Could you let us know why you are trying to install elasticsearch 8.12.

Thank you.

**Comment by hostep** [comment]
Okay, sure fine, your argument makes no sense but I won't put my energy into trying to convince you otherwise.

But let's go with your argument and say you want to install the elasticsearch php package version 8.7.x, that is also not going to work, because of the constraint `~8.5.0` ...

**Comment by engcom-November** [comment]
@hostep, I agree with you, the elasticsearch constraint should be changed to `~8.7.0`, marking this as confirmed.

Thank you.

**Comment by hostep** [comment]
@engcom-November: the constraint should be changed to `^8.0` or `^8.5` or `^8.7` (doesn't really matter). Please don't force your users to use old versions of Elasticsearch or their associated php library.

Elasticsearch has brought out a bunch of security fixes in for example versions 8.8, 8.9, 8.10 (see [here](https://www.elastic.co/community/security), scroll down a bit until you reached Elasticsearch), why do you want to force your users to stay on old vulnerable versions of Elasticsearch? Please don't....

Also your 'system requirements' page only shows the versions of software that are guaranteed to work, but those aren't the strict requirements to run Magento, other versions are also compatible, they just haven't been verified by Adobe to work correctly, but please leave that up to the community to deal with, don't force us to use old software with potential security issues.

**Comment by hostep** [comment]
Another argument, is that the so called ["system requirements"](https://experienceleague.adobe.com/docs/commerce-operations/installation-guide/system-requirements.html?lang=en) were updated yesterday for Magento version 2.4.6-p4 and increased supported Elasticsearch version to 8.11
I still believe the php composer dependency should allow a more wider range of versions and switching from `~` to `^` constraint would be better.

**Comment by hostep** [comment]
Closing, this was fixed in https://github.com/magento/magento2/commit/71bfd9a0c7442424b5727038fbaa69d319330146 to be released on Magento 2.4.8 if I got it right.
